### PR TITLE
#107 Add search item index and sort items

### LINKED ISSUE #107: Add search item index and sort items.
## Kind (Required)

- Improvement

 ## Overview (Required)

- Add search item index of speaker and session.
  - The same index is displayed only once. 
  - The index is displayed in the same way as the time in the timetable.
- Sort by speaker name, session title.

 ## Links	

-

### PR DESCRIPTION
**Pull request by Pluu**
## Issue
- close #107

## Overview (Required)
- The basics of implementing ItemDecoration were based on DroidKaigi 2019
- X/Y position other than the "textAppearanceHeader6" specification may differ from the design guide.

## Links
- [SearchFragment#StickyHeaderItemDecoration](https://github.com/DroidKaigi/conference-app-2019/blob/master/feature/session/src/main/java/io/github/droidkaigi/confsched2019/session/ui/SearchFragment.kt#L237)

## Screenshot
Before | After
:--: | :--:
 |

### CODE REVIEW

**Comment by jmatsu** [comment]
👀

File: feature/session/src/main/java/io/github/droidkaigi/confsched2020/session/ui/widget/SearchItemDecoration.kt [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import androidx.core.content.res.ResourcesCompat
+import androidx.core.content.res.getColorOrThrow
+import androidx.core.content.res.getDimensionOrThrow
+import androidx.core.content.res.getResourceIdOrThrow
+import androidx.recyclerview.widget.RecyclerView
+import io.github.droidkaigi.confsched2020.session.R
+
+class SearchItemDecoration(
+    context: Context,
+    private val getGroupId: (Int) -> Long,
```
[the remark this answers is not in the document]
Comment (Pluu) [reply feature/session/src/main/java/io/github/droidkaigi/confsched2020/session/ui/widget/SearchItemDecoration.kt]: Good! implementation is clearer!

File: feature/session/src/main/java/io/github/droidkaigi/confsched2020/session/ui/widget/SearchItemDecoration.kt [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    ) {
+        super.onDraw(c, parent, state)
+
+        val totalItemCount = state.itemCount
+        val childCount = parent.childCount
+        val lineHeight = textPaint.textSize + fontMetrics.descent
+        var previousGroupId: Long
+        var groupId: Long = EMPTY_ID
+
+        (0 until childCount).forEach {
```
[the remark this answers is not in the document]
Comment (Pluu) [reply feature/session/src/main/java/io/github/droidkaigi/confsched2020/session/ui/widget/SearchItemDecoration.kt]: Good Point!

File: feature/session/src/main/java/io/github/droidkaigi/confsched2020/session/ui/SearchSessionsFragment.kt [hunk 3]
```diff
@@ -97,13 +118,17 @@ class SearchSessionsFragment : DaggerFragment() {
                 groupAdapter.add(sectionHeaderItemFactory.create(resources.getString(R.string.speaker)))
                 groupAdapter.addAll(uiModel.searchResult.speakers.map {
                     speakerItemFactory.create(it)
+                }.sortedBy {
+                    it.speaker.name.toUpperCase(Locale.getDefault())
```
[the remark this answers is not in the document]
Comment (Pluu) [reply feature/session/src/main/java/io/github/droidkaigi/confsched2020/session/ui/SearchSessionsFragment.kt]: Currently, speakers use both names and nickname.
If you use the suggestions, "A ~~~" and "a ~~~" will be exposed in different locations.
Last year they were exposed to the same location.

This requires confirmation by the organizer.
cc @takahirom

[the remark this answers is not in the document]
Comment (jmatsu) [reply feature/session/src/main/java/io/github/droidkaigi/confsched2020/session/ui/SearchSessionsFragment.kt]: Ah got it. make sense. The current implementation is nicer! Thanks @Pluu

**Comment by takahirom** [comment]
Thanks!

**Pull request merged**